src/workspace.ts

```diff
@@ -83,4 +83,36 @@ export class MonacoWorkspace implements Workspace {
         return this.onDidChangeTextDocumentEmitter.event;
     }
 
+    public applyEdit(workspaceEdit: WorkspaceEdit): Thenable<boolean> {
+        let applied = true;
+        if (workspaceEdit.documentChanges) {
+            for (const change of workspaceEdit.documentChanges) {
+                if (change.textDocument.version && change.textDocument.version >= 0) {
+                    const textDocument = this.documents.get(change.textDocument.uri);
+                    if (textDocument && textDocument.version === change.textDocument.version) {
+                        monaco.editor.getModel(monaco.Uri.parse(textDocument.uri)).pushEditOperations(
+                            [],  // Do not try and preserve editor selections.
+                            change.edits.map((edit: TextEdit) => {
+                                return {
+                                    identifier: {major: 1, minor: 0},
+                                    range: this.p2m.asRange(edit.range),
+                                    text: edit.newText,
+                                    forceMoveMarkers: true,
+                                };
+                            }),
+                            () => [],  // Do not try and preserve editor selections.
+                        );
+                    } else {
+                        applied = false;
+                    }
+                } else {
+                    applied = false;
```

Comment:
Is not the whole change should be applied or rejected? It seems that changes can be applied partially by this method.